src/main/java/tictactoe/Minimax.java

```diff
@@ -0,0 +1,81 @@
+package tictactoe;
+
+import java.util.HashMap;
+import java.util.Map;
+
+public class Minimax {
+
+    public int chooseMove
+            (Board board, String currentPlayersMark, String opponentsMark) {
+
+        var copyBoard = board.copyBoard();
+        return findBestMove(copyBoard, 0, currentPlayersMark, opponentsMark);
+    }
+
+    private int findBestMove
+            (Board board, int depth, String currentPlayersMark, String opponentsMark) {
+        var movesAndScores = new HashMap<Integer, Integer>();
+        var availableSquares = board.availableSquares(currentPlayersMark, opponentsMark);
+
+        if (board.isComplete(currentPlayersMark, opponentsMark)) {
+            return scoreMove(board, depth, currentPlayersMark, opponentsMark);
+        }
+
+        for (String square : availableSquares) {
+            var squareNum = Integer.parseInt(square);
+            var copyBoard = board.copyBoard();
+
+            copyBoard.takeSquare(squareNum, currentPlayersMark);
+
+            movesAndScores.put(squareNum,
+                    -1 * findBestMove(copyBoard, depth + 1, opponentsMark, currentPlayersMark));
+        }
+        return evaluateMove(depth, movesAndScores);
+    }
+
+
+    public int maximisingPlayerBestMove(Map<Integer, Integer> scores) {
```

Comment:
I think you can refactor this so that you don't need to access the first item from the iterator for the entry set just to get an initial value. Can you think of some value you could assign to `bestScore` so that the first iteration of the *for each* loop will assign a sane value to it?

Reply:
Separate but related point: I think the way you're using the Map is really to create a list of pairs of values: as there's no real key-value relationship between the data. Whilst it's more verbose, I'd be tempted to have a simple (inner) class:
```
class Result {
  private final int move;
  private final int score;

  public Result(int move, int score) {
   this.move = move;
   this.score = score;
  }

  ... getters and setters
}
```

You could then just use a type `List<Result>`.

(FYI - this is what people are talking about when they say Java is a verbose language :))

Reply:
> I think you can refactor this so that you don't need to access the first item from the iterator for the entry set just to get an initial value. Can you think of some value you could assign to `bestScore` so that the first iteration of the _for each_ loop will assign a sane value to it?

Good point. The lowest score the maximising player can get is 0, which can be assigned to `bestScore`

Reply:
> Separate but related point: I think the way you're using the Map is really to create a list of pairs of values: as there's no real key-value relationship between the data. Whilst it's more verbose, I'd be tempted to have a simple (inner) class:
> 
> ```
> class Result {
>   private final int move;
>   private final int score;
> 
>   public Result(int move, int score) {
>    this.move = move;
>    this.score = score;
>   }
> 
>   ... getters and setters
> }
> ```
> 
> You could then just use a type `List<Result>`.
> 
> (FYI - this is what people are talking about when they say Java is a verbose language :))

Thanks, @matthewglover. Creating a separate class to handle the results sounds great. I'm a bit confused with how to replace the current hashmap with a list containing results. Is this something we can go through on Wednesday?